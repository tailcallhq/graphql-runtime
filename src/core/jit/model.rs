use std::collections::HashMap;
use std::fmt::{Debug, Formatter};

use async_graphql::parser::types::OperationType;
use async_graphql::Pos;
use serde::Deserialize;

use crate::core::ir::model::IR;

#[derive(Debug, Deserialize, Clone)]
pub struct Variables<Value>(HashMap<String, Value>);

impl<Value> Default for Variables<Value> {
    fn default() -> Self {
        Self::new()
    }
}

impl<Value> Variables<Value> {
    pub fn new() -> Self {
        Self(HashMap::new())
    }
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.0.get(key)
    }
    pub fn insert(&mut self, key: String, value: Value) {
        self.0.insert(key, value);
    }
}

impl<V> FromIterator<(String, V)> for Variables<V> {
    fn from_iter<T: IntoIterator<Item = (String, V)>>(iter: T) -> Self {
        Self(iter.into_iter().collect())
    }
}

#[derive(Debug, Clone)]
pub struct Arg<Input> {
    pub id: ArgId,
    pub name: String,
    pub type_of: crate::core::blueprint::Type,
    pub value: Option<Input>,
    pub default_value: Option<Input>,
}

impl<Input> Arg<Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Arg<Output>, Error> {
        Ok(Arg {
            id: self.id,
            name: self.name,
            type_of: self.type_of,
            value: self.value.map(&map).transpose()?,
            default_value: self.default_value.map(&map).transpose()?,
        })
    }
}

#[derive(Clone)]
pub struct ArgId(usize);

impl Debug for ArgId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl ArgId {
    pub fn new(id: usize) -> Self {
        ArgId(id)
    }
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct FieldId(usize);

impl Debug for FieldId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl FieldId {
    pub fn new(id: usize) -> Self {
        FieldId(id)
    }
    pub fn as_usize(&self) -> usize {
        self.0
    }
}

#[derive(Clone)]
pub struct Field<Extensions, Input> {
    pub id: FieldId,
    pub name: String,
    pub ir: Option<IR>,
    pub type_of: crate::core::blueprint::Type,
    pub skip: Option<Variable>,
    pub include: Option<Variable>,
    pub args: Vec<Arg<Input>>,
    pub extensions: Option<Extensions>,
    pub pos: Pos,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Variable(String);

impl Variable {
    pub fn new(name: String) -> Self {
        Variable(name)
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
    pub fn into_string(self) -> String {
        self.0
    }
}

impl<Extensions, Input> Field<Extensions, Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Field<Extensions, Output>, Error> {
        Ok(Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            extensions: self.extensions,
            skip: self.skip,
            include: self.include,
            pos: self.pos,
            args: self
                .args
                .into_iter()
                .map(|arg| arg.try_map(&map))
                .collect::<Result<_, _>>()?,
        })
    }
}

impl<Input> Field<Nested<Input>, Input> {
    pub fn nested(&self) -> Option<&Vec<Field<Nested<Input>, Input>>> {
        self.extensions.as_ref().map(|Nested(nested)| nested)
    }

    pub fn nested_iter(&self) -> impl Iterator<Item = &Field<Nested<Input>, Input>> {
        self.nested()
            .map(|nested| nested.iter())
            .into_iter()
            .flatten()
    }
}

impl<Input> Field<Flat, Input> {
    fn parent(&self) -> Option<&FieldId> {
        self.extensions.as_ref().map(|Flat(id)| id)
    }

    fn into_nested(self, fields: &[Field<Flat, Input>]) -> Field<Nested<Input>, Input>
    where
        Input: Clone,
    {
        let mut children = Vec::new();
        for field in fields.iter() {
            if let Some(id) = field.parent() {
                if *id == self.id {
                    children.push(field.to_owned().into_nested(fields));
                }
            }
        }

        let extensions = if children.is_empty() {
            None
        } else {
            Some(Nested(children))
        };

        Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            skip: self.skip,
            include: self.include,
            args: self.args,
            pos: self.pos,
            extensions,
        }
    }
}

impl<Extensions: Debug, Input: Debug> Debug for Field<Extensions, Input> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut debug_struct = f.debug_struct("Field");
        debug_struct.field("id", &self.id);
        debug_struct.field("name", &self.name);
        if self.ir.is_some() {
            debug_struct.field("ir", &"Some(..)");
        }
        debug_struct.field("type_of", &self.type_of);
        if !self.args.is_empty() {
            debug_struct.field("args", &self.args);
        }
        if self.extensions.is_some() {
            debug_struct.field("extensions", &self.extensions);
        }
        debug_struct.finish()
    }
}

/// Stores field relationships in a flat structure where each field links to its
/// parent.
#[derive(Clone)]
pub struct Flat(FieldId);

impl Flat {
    pub fn new(id: FieldId) -> Self {
        Flat(id)
    }
    pub fn as_id(&self) -> &FieldId {
        &self.0
    }
}
impl Debug for Flat {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Flat({:?})", self.0)
    }
}

/// Store field relationships in a nested structure like a tree where each field
/// links to its children.
#[derive(Clone, Debug)]
pub struct Nested<Input>(Vec<Field<Nested<Input>, Input>>);

#[derive(Clone, Debug)]
pub struct OperationPlan<Input> {
    flat: Vec<Field<Flat, Input>>,
    operation_type: OperationType,
    nested: Vec<Field<Nested<Input>, Input>>,
}

impl<Input> OperationPlan<Input> {
    pub fn new(fields: Vec<Field<Flat, Input>>, operation_type: OperationType) -> Self
    where
        Input: Clone,
    {
        let nested = fields
            .clone()
            .into_iter()
            .filter(|f| f.extensions.is_none())
            .map(|f| f.into_nested(&fields))
            .collect::<Vec<_>>();

        Self { flat: fields, nested, operation_type }
    }

    pub fn operation_type(&self) -> OperationType {
        self.operation_type
    }

    pub fn is_query(&self) -> bool {
        self.operation_type == OperationType::Query
    }

    pub fn as_nested(&self) -> &[Field<Nested<Input>, Input>] {
        &self.nested
    }

    pub fn into_nested(self) -> Vec<Field<Nested<Input>, Input>> {
        self.nested
    }

    pub fn as_parent(&self) -> &[Field<Flat, Input>] {
        &self.flat
    }

    pub fn find_field(&self, id: FieldId) -> Option<&Field<Flat, Input>> {
        self.flat.iter().find(|field| field.id == id)
    }

    pub fn find_field_path<S: AsRef<str>>(&self, path: &[S]) -> Option<&Field<Flat, Input>> {
        match path.split_first() {
            None => None,
            Some((name, path)) => {
                let field = self.flat.iter().find(|field| field.name == name.as_ref())?;
                if path.is_empty() {
                    Some(field)
                } else {
                    self.find_field_path(path)
                }
            }
        }
    }

    pub fn size(&self) -> usize {
        self.flat.len()
    }
}
