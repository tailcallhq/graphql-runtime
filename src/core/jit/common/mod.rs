mod json_placeholder;
pub use json_placeholder::*;
