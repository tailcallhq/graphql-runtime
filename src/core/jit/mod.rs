mod builder;
mod eval_ir;
mod model;
mod store;
mod synth;
pub use builder::*;
pub use eval_ir::*;
pub use model::*;
pub use store::*;
pub use synth::*;

// NOTE: Only used in tests and benchmarks
pub mod common;
pub mod ir;
