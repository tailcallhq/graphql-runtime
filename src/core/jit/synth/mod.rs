// mod synth_borrow;
mod synth_const;

// pub use synth_borrow::SynthBorrow;
pub use synth_const::{Synth, SynthConst};

use super::Store;
pub trait Synthesizer {
    type Value;
    fn synthesize(self, store: Store<Self::Value>) -> Self::Value;
}
