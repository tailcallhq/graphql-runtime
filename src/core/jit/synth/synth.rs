use std::borrow::Cow;

use crate::core::jit::model::{Field, OperationPlan, Variables};
use crate::core::jit::store::{DataPath, Store};
use crate::core::jit::{Error, PathSegment, Positioned, ValidationError};
use crate::core::json::{JsonLike, JsonObjectLike};

type ValueStore<Value> = Store<Result<Value, Positioned<Error>>>;

pub struct Synth<'a, Value> {
    plan: &'a OperationPlan<Value>,
    store: ValueStore<Value>,
    variables: Variables<Value>,
}

impl<'a, Value> Synth<'a, Value> {
    #[inline(always)]
    pub fn new(
        plan: &'a OperationPlan<Value>,
        store: ValueStore<Value>,
        variables: Variables<Value>,
    ) -> Self {
        Self { plan, store, variables }
    }
}

impl<'a, Value> Synth<'a, Value>
where
    Value: JsonLike<'a> + Clone + std::fmt::Debug,
{
    #[inline(always)]
    fn include(&self, field: &Field<Value>) -> bool {
        !field.skip(&self.variables)
    }

    #[inline(always)]
    pub fn synthesize(&'a self) -> Result<Value, Positioned<Error>> {
        let mut data = Value::JsonObject::with_capacity(self.plan.selection.len());
        let mut path = Vec::new();
        let root_name = self.plan.root_name();

        for child in self.plan.selection.iter() {
            if !self.include(child) {
                continue;
            }
            // TODO: in case of error set `child.output_name` to null
            // and append error to response error array
            let val = self.iter(child, None, &DataPath::new(), &mut path, Some(root_name))?;
            data.insert_key(&child.output_name, val);
        }

        Ok(Value::object(data))
    }

    #[allow(clippy::too_many_arguments)]
    #[inline(always)]
    fn iter(
        &'a self,
        node: &'a Field<Value>,
        value: Option<&'a Value>,
        data_path: &DataPath,
        path: &mut Vec<PathSegment>,
        root_name: Option<&'a str>,
    ) -> Result<Value, Positioned<Error>> {
        path.push(PathSegment::Field(node.output_name.clone()));

        let result = match self.store.get(&node.id) {
            Some(value) => {
                let mut value = value.as_ref().map_err(Clone::clone)?;

                for index in data_path.as_slice() {
                    if let Some(arr) = value.as_array() {
                        value = &arr[*index];
                    } else {
                        return Ok(Value::null());
                    }
                }

                if node.type_of.is_list() != value.as_array().is_some() {
                    return self.node_nullable_guard(node, path, None);
                }
                self.iter_inner(node, value, data_path, path)
            }
            None => match value {
                Some(result) => self.iter_inner(node, result, data_path, path),
                None => self.node_nullable_guard(node, path, root_name),
            },
        };

        path.pop();
        result
    }

    /// This guard ensures to return Null value only if node type permits it, in
    /// case it does not it throws an Error
    fn node_nullable_guard(
        &'a self,
        node: &'a Field<Value>,
        path: &[PathSegment],
        root_name: Option<&'a str>,
    ) -> Result<Value, Positioned<Error>> {
        if let Some(root_name) = root_name {
            if node.name.eq("__typename") {
                return Ok(Value::string(Cow::Borrowed(root_name)));
            }
        }
        // according to GraphQL spec https://spec.graphql.org/October2021/#sec-Handling-Field-Errors
        if node.type_of.is_nullable() {
            Ok(Value::null())
        } else {
            Err(ValidationError::ValueRequired.into())
                .map_err(|e| self.to_location_error(e, node, path))
        }
    }

    #[allow(clippy::too_many_arguments)]
    #[inline(always)]
    fn iter_inner(
        &'a self,
        node: &'a Field<Value>,
        value: &'a Value,
        data_path: &DataPath,
        path: &mut Vec<PathSegment>,
    ) -> Result<Value, Positioned<Error>> {
        // skip the field if field is not included in schema
        if !self.include(node) {
            return Ok(Value::null());
        }

        let eval_result = if value.is_null() {
            // check the nullability of this type unwrapping list modifier
            let is_nullable = match &node.type_of {
                crate::core::Type::Named { non_null, .. } => !*non_null,
                crate::core::Type::List { of_type, .. } => of_type.is_nullable(),
            };
            if is_nullable {
                Ok(Value::null())
            } else {
                Err(ValidationError::ValueRequired.into())
            }
        } else if node.scalar.is_some() {
            let scalar = node.scalar.as_ref().unwrap();

            // TODO: add validation for input type as well. But input types are not checked
            // by async_graphql anyway so it should be done after replacing
            // default engine with JIT
            if scalar.validate(value) {
                Ok(value.clone())
            } else {
                Err(
                    ValidationError::ScalarInvalid { type_of: node.type_of.name().to_string() }
                        .into(),
                )
            }
        } else if node.is_enum {
            let check_valid_enum = |value: &Value| -> bool {
                value
                    .as_str()
                    .map(|v| self.plan.field_validate_enum_value(node, v))
                    .unwrap_or(false)
            };

            let is_valid_enum = if let Some(vec) = value.as_array() {
                vec.iter().all(check_valid_enum)
            } else {
                check_valid_enum(value)
            };

            if is_valid_enum {
                Ok(value.clone())
            } else {
                Err(
                    ValidationError::EnumInvalid { type_of: node.type_of.name().to_string() }
                        .into(),
                )
            }
        } else {
            match (value.as_array(), value.as_object()) {
                (_, Some(obj)) => {
                    let mut ans = Value::JsonObject::with_capacity(node.selection.len());

                    for child in node
                        .iter()
                        .filter(|field| self.plan.field_is_part_of_value(field, value))
                    {
                        // all checks for skip must occur in `iter_inner`
                        // and include be checked before calling `iter` or recursing.
                        if self.include(child) {
                            let value = if child.name == "__typename" {
                                Value::string(node.value_type(value).into())
                            } else {
                                let val = obj.get_key(child.name.as_str());
                                self.iter(child, val, data_path, path, None)?
                            };
                            ans.insert_key(&child.output_name, value);
                        }
                    }

                    Ok(Value::object(ans))
                }
                (Some(arr), _) => {
                    let mut ans = Vec::with_capacity(arr.len());
                    for (i, val) in arr.iter().enumerate() {
                        path.push(PathSegment::Index(i));
                        let val =
                            self.iter_inner(node, val, &data_path.clone().with_index(i), path)?;
                        path.pop();
                        ans.push(val);
                    }
                    Ok(Value::array(ans))
                }
                _ => Ok(value.clone()),
            }
        };

        eval_result.map_err(|e| self.to_location_error(e, node, path))
    }

    fn to_location_error(
        &'a self,
        error: Error,
        node: &'a Field<Value>,
        path: &[PathSegment],
    ) -> Positioned<Error> {
        Positioned::new(error, node.pos).with_path(path.to_vec())
    }
}

#[cfg(test)]
mod tests {
    use async_graphql_value::ConstValue;
    use serde::{Deserialize, Serialize};

    use super::ValueStore;
    use crate::core::blueprint::Blueprint;
    use crate::core::config::{Config, ConfigModule};
    use crate::core::jit::builder::Builder;
    use crate::core::jit::fixtures::JP;
    use crate::core::jit::model::{FieldId, Variables};
    use crate::core::jit::store::Store;
    use crate::core::jit::synth::Synth;
    use crate::core::jit::OperationPlan;
    use crate::core::json::JsonLike;
    use crate::core::valid::Validator;

    const POSTS: &str = r#"
        [
                {
                    "id": 1,
                    "userId": 1,
                    "title": "Some Title"
                },
                {
                    "id": 2,
                    "userId": 1,
                    "title": "Not Some Title"
                }
        ]
    "#;

    const USER1: &str = r#"
        {
                "id": 1,
                "name": "foo"
        }
    "#;

    const USER2: &str = r#"
        {
                "id": 2,
                "name": "bar"
        }
    "#;
    const USERS: &str = r#"
        [
          {
            "id": 1,
            "name": "Leanne Graham"
          },
          {
            "id": 2,
            "name": "Ervin Howell"
          }
        ]
    "#;

    #[derive(Clone)]
    enum TestData {
        Posts,
        UsersData,
        Users,
        User1,
    }

    impl TestData {
        fn into_value<'a, Value: Deserialize<'a> + JsonLike<'a>>(self) -> Value {
            match self {
                Self::Posts => serde_json::from_str(POSTS).unwrap(),
                Self::User1 => serde_json::from_str(USER1).unwrap(),
                TestData::UsersData => Value::array(vec![
                    serde_json::from_str(USER1).unwrap(),
                    serde_json::from_str(USER2).unwrap(),
                ]),
                TestData::Users => serde_json::from_str(USERS).unwrap(),
            }
        }
    }

    const CONFIG: &str = include_str!("../fixtures/jsonplaceholder-mutation.graphql");

    fn make_store<'a, Value>(
        query: &str,
        store: Vec<(FieldId, TestData)>,
    ) -> (OperationPlan<Value>, ValueStore<Value>, Variables<Value>)
    where
        Value: Deserialize<'a> + JsonLike<'a> + Serialize + Clone + std::fmt::Debug,
    {
        let store = store
            .into_iter()
            .map(|(id, data)| (id, data.into_value()))
            .collect::<Vec<_>>();

        let doc = async_graphql::parser::parse_query(query).unwrap();
        let config = Config::from_sdl(CONFIG).to_result().unwrap();
        let config = ConfigModule::from(config);

        let builder = Builder::new(&Blueprint::try_from(&config).unwrap(), doc);
        let plan = builder.build(None).unwrap();
        let plan = plan
            .try_map(|v| {
                // Earlier we hard OperationPlan<ConstValue> which has impl Deserialize
                // but now InputResolver takes OperationPlan<async_graphql_value::Value>
                // and returns OperationPlan<async_graphql_value::Value>.
                // So we need to map Plan to some other value before being able to deserialize
                // it.
                let serde = v.into_json().unwrap();
                Deserialize::deserialize(serde)
            })
            .unwrap();

        let store = store
            .into_iter()
            .fold(Store::new(), |mut store, (id, data)| {
                store.set_data(id, Ok(data));
                store
            });
        let vars = Variables::new();

        (plan, store, vars)
    }

    fn assert_synths(query: &str, store: Vec<(FieldId, TestData)>) {
        let (plan, value_store, vars) = make_store::<ConstValue>(query, store.clone());
        let synth_const = Synth::new(&plan, value_store, vars);
        let (plan, value_store, vars) =
            make_store::<serde_json_borrow::Value>(query, store.clone());
        let synth_borrow = Synth::new(&plan, value_store, vars);

        let val_const = synth_const.synthesize().unwrap();
        let val_const = serde_json::to_string_pretty(&val_const).unwrap();
        let val_borrow = synth_borrow.synthesize().unwrap();
        let val_borrow = serde_json::to_string_pretty(&val_borrow).unwrap();
        assert_eq!(val_const, val_borrow);
    }

    #[test]
    fn test_posts() {
        let store = vec![(FieldId::new(0), TestData::Posts)];
        let query = r#"
            query {
                posts { id }
            }
        "#;

        assert_synths(query, store);
    }

    #[test]
    fn test_user() {
        let store = vec![(FieldId::new(0), TestData::User1)];
        let query = r#"
                query {
                    user(id: 1) { id }
                }
            "#;

        assert_synths(query, store);
    }

    #[test]
    fn test_nested() {
        let store = vec![
            (FieldId::new(0), TestData::Posts),
            (FieldId::new(3), TestData::UsersData),
        ];
        let query = r#"
                query {
                    posts { id title user { id name } }
                }
            "#;
        assert_synths(query, store);
    }

    #[test]
    fn test_multiple_nested() {
        let store = vec![
            (FieldId::new(0), TestData::Posts),
            (FieldId::new(3), TestData::UsersData),
            (FieldId::new(6), TestData::Users),
        ];
        let query = r#"
                query {
                    posts { id title user { id name } }
                    users { id name }
                }
            "#;
        assert_synths(query, store);
    }

    #[test]
    fn test_json_placeholder() {
        let jp = JP::init("{ posts { id title userId user { id name } } }", None);
        let synth = jp.synth();
        let val: async_graphql::Value = synth.synthesize().unwrap();
        insta::assert_snapshot!(serde_json::to_string_pretty(&val).unwrap())
    }

    #[test]
    fn test_json_placeholder_borrowed() {
        let jp = JP::init("{ posts { id title userId user { id name } } }", None);
        let synth = jp.synth();
        let val: serde_json_borrow::Value = synth.synthesize().unwrap();
        insta::assert_snapshot!(serde_json::to_string_pretty(&val).unwrap())
    }

    #[test]
    fn test_json_placeholder_typename() {
        let jp = JP::init("{ posts { id __typename user { __typename id } } }", None);
        let synth = jp.synth();
        let val: serde_json_borrow::Value = synth.synthesize().unwrap();
        insta::assert_snapshot!(serde_json::to_string_pretty(&val).unwrap())
    }

    #[test]
    fn test_json_placeholder_typename_root_level() {
        let jp = JP::init("{ __typename posts { id user { id }} }", None);
        let synth = jp.synth();
        let val: serde_json_borrow::Value = synth.synthesize().unwrap();
        insta::assert_snapshot!(serde_json::to_string_pretty(&val).unwrap())
    }
}
