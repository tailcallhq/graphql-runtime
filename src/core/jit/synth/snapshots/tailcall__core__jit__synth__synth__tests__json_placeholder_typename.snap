---
source: src/core/jit/synth/synth.rs
expression: "serde_json::to_string_pretty(&val).unwrap()"
---
{
  "posts": [
    {
      "id": 1,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 2,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 3,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 4,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 5,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 6,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 7,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 8,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 9,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 10,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 1
      }
    },
    {
      "id": 11,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 12,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 13,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 14,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 15,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 16,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 17,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 18,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 19,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 20,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 2
      }
    },
    {
      "id": 21,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 22,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 23,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 24,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 25,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 26,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 27,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 28,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 29,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 30,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 3
      }
    },
    {
      "id": 31,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 32,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 33,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 34,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 35,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 36,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 37,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 38,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 39,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 40,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 4
      }
    },
    {
      "id": 41,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 42,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 43,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 44,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 45,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 46,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 47,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 48,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 49,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 50,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 5
      }
    },
    {
      "id": 51,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 52,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 53,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 54,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 55,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 56,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 57,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 58,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 59,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 60,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 6
      }
    },
    {
      "id": 61,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 62,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 63,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 64,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 65,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 66,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 67,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 68,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 69,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 70,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 7
      }
    },
    {
      "id": 71,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 72,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 73,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 74,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 75,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 76,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 77,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 78,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 79,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 80,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 8
      }
    },
    {
      "id": 81,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 82,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 83,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 84,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 85,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 86,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 87,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 88,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 89,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 90,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 9
      }
    },
    {
      "id": 91,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 92,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 93,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 94,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 95,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 96,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 97,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 98,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 99,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    },
    {
      "id": 100,
      "__typename": "Post",
      "user": {
        "__typename": "User",
        "id": 10
      }
    }
  ]
}
